Instructor Dashboard: Add functionality for obtaining list of students who may enroll in a course but haven't signed up yet

itsjeyd opened the pull request:
## Background

This PR addresses JIRA ticket [TNL-1652](https://openedx.atlassian.net/browse/TNL-1652), which is based on the following feature proposal: [OSPR-171](https://openedx.atlassian.net/browse/OSPR-171)

It moves functionality for obtaining a list of students who may enroll in a given course but have not signed up yet from the legacy dashboard to the Instructor Dashboard.

The functionality that this PR adds to the Instructor Dashboard differs from the functionality that used to be available on the legacy dashboard in two ways:

1.  Instead of displaying a list of students who may enroll in a course on the dashboard, results are made available for download in the form of a CSV file.
2.  Existing enrollments are taken into account when computing the list of students who may enroll in a course. This means that students who are allowed to enroll in the course **and** have already done so will not be added to the resulting CSV file.

## Affected components

This PR affects the LMS. In particular, it introduces changes to the legacy dashboard and the new instructor dashboard (as explained above).

## Affected users

The changes introduced in this PR only affect users with staff privileges; they do not affect all edX students.

## Instructions for manual testing

1.  On Devstack, start a Django shell via

        ./manage.py lms shell --settings=devstack

    and add test data by executing the following code:

        from student.models import CourseEnrollmentAllowed
        course_id = "edX/DemoX/Demo_Course"
        for n in range(3):
            CourseEnrollmentAllowed.objects.create(email="student{}@example.com".format(n), course_id=course_id)

        CourseEnrollmentAllowed.objects.create(email="staff@example.com", course_id=course_id)

2.  Run the LMS and sign in as `staff@example.com`.

3.  Navigate to <http://localhost:8000/courses/edX/DemoX/Demo_Course/instructor> and click on the "Data Download" tab.

4.  Click "Download info about students who may enroll as a CSV" to generate a CSV file with the same data (will be added to the list of files available for download when it's ready).

5.  Check the contents of the CSV file (as `vagrant` user) via:

        cd /tmp/edx-s3/grades/edX%2FDemoX%2FDemo_Course/
        cat edX_DemoX_Demo_Course_may_enroll_info_YYYY-MM-DD-HHMM.csv

Observe that the results do not include students who are already enrolled in the Demo Course (in this case: `staff@example.com`).

## Screenshots

### Legacy dashboard

Legacy dashboard displaying instructions on where to find deprecated functionality:

![legacy-dashboard-message](https://cloud.githubusercontent.com/assets/961441/7771810/604e294c-009a-11e5-9d6b-c7903779b31f.png)

### Instructor dashboard

New controls for downloading list of students who may enroll:

![data-download-reports-highlighted](https://cloud.githubusercontent.com/assets/961441/7771828/72a64e58-009a-11e5-8d75-534a03444019.png)

CSV generation in progress:

![data-download-reports-in-progress-highlighted](https://cloud.githubusercontent.com/assets/961441/7771844/8309b65e-009a-11e5-8b12-5ac747e7ae65.png)

Results have become available:

![data-download-reports-with-results-highlighted](https://cloud.githubusercontent.com/assets/961441/7771855/8f11160e-009a-11e5-9e40-6bba36bfe10d.png)

## Licensing

This PR is covered by the OpenCraft contributor agreement (@antoviaque).

---

Please review this PR. I'll be happy to incorporate any suggestions for improvements. Thanks!

edx-webhook commented:
Thanks for the pull request, @itsjeyd! I've created [OSPR-607](https://openedx.atlassian.net/browse/OSPR-607) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've submitted a [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf) or indicated your institutional affiliation and added yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

sarina commented:
add to whitelist

sarina commented on common/djangoapps/student/models.py:
```diff
@@ -1293,6 +1293,32 @@ class Meta:  # pylint: disable=missing-docstring
+        return CourseEnrollmentAllowed.objects.filter(course_id=course_id)
+
+    @classmethod
+    def num_may_enroll_in(cls, course_id):
```
Given the name of this method and the name of the above method, I would expect this method to return the number of the last method (ie `len(may_enroll_in)`). However that's not the case, which is really confusing. I will read further to try to figure out how you're using these and make suggestions, but this can't stay as-is.

itsjeyd replied:
@sarina I agree. Would renaming it to `num_may_enroll_but_unenrolled` make it clearer? I only use this method once throughout the rest of the code, but I still feel like it makes sense to encapsulate the logic for computing this number in the model class. What do you think?

sarina commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -1742,6 +1742,11 @@ def setUp(self):
+        self.students_who_may_enroll = self.students + [UserFactory() for _ in xrange(6, 11)]
+        for student in self.students_who_may_enroll:
+            CourseEnrollmentAllowed.objects.create(
+                email=student.email, course_id=self.course.id)
```
Closing paren should always be on the next line (you have a lot of this throughout the PR you'll need to fix up). Please see https://github.com/edx/edx-platform/wiki/Python-Guidelines#breaking-long-lines

sarina commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -1742,6 +1742,11 @@ def setUp(self):
         for student in self.students:
             CourseEnrollment.enroll(student, self.course.id)
 
+        self.students_who_may_enroll = self.students + [UserFactory() for _ in xrange(6, 11)]
```
Why are you doing xrange(6,11)? range(5) or xrange(5) would be clearer.

sarina commented:
There are some tests you'll want to add in test_api.py. At the beginning of the file there's some data for reports download ("# ddt data for test cases involving reports") - if you can add data for this download into that, great.

Also in `class TestInstructorAPIDenyLevels` there's a list of endpoints - you should add this endpoint to the proper list to make sure it's included in those tests.

sarina commented:
Question for @explorerleslie : what should be done when the report is empty? Currently, there's just an empty report with a header row "email" but no data. Should something be explicitly added to the report to indicate that no students meet the criteria?

sarina commented:
@mhoeber - this will need both documentation of the feature, as well as a doc review of newly added strings.

explorerleslie commented:
@sarina nah. I don't think we do that for any other report. Empty is fine.

itsjeyd commented:
@sarina The latest commits should address your comments.

I changed the API of the `CourseEnrollmentAllowed` class to make it more consistent. I renamed the class methods to `may_enroll_and_unenrolled` and `num_may_enroll_and_unenrolled`, respectively. They do what their names suggest: Return a list/the number of students who may enroll and are currently unenrolled.

sarina commented:
If you click to your CI build page, then on "Build Graph", you can see your build is failing because of quality. To check quality locally, run `paver run_quality`

I have a sandbox set up for this. I enrolled 4 users via the Membership tab, then generated a report - the report came back empty.

sarina commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -865,6 +865,38 @@ def upload_students_csv(_xmodule_instance_args, _entry_id, course_id, task_input
+    """
+    start_time = time()
+    start_date = datetime.now(UTC)
+    num_students = CourseEnrollmentAllowed.num_may_enroll_and_unenrolled(course_id)
```
I'm not certain this makes sense. This does the calculation we want in advance of doing the calculation we want, all to obtain a number for the task progress.... sorry now I'm dizzy :smile: 

But yeah - I'm not certain this is the best use of resources.

sarina commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -865,6 +865,38 @@ def upload_students_csv(_xmodule_instance_args, _entry_id, course_id, task_input
+    start_time = time()
+    start_date = datetime.now(UTC)
+    num_students = CourseEnrollmentAllowed.num_may_enroll_and_unenrolled(course_id)
+    task_progress = TaskProgress(action_name, num_students, start_time)
```
I wonder: there's nothing that i can see in the definition of `TaskProgress` that requires the second argument to actually be a number. I wonder if we avoid doing this calculation twice, and instead just insert some string here (eg "all students") or just insert the number 1 (to indicate we're generating 1 report). Frankly this generation will in the majority of cases happen so fast people probably won't even see the Task Progress bar get updated.

itsjeyd commented:
@sarina Thanks for pointing out why the build was failing. Looking into the second issue (empty reports) now.

itsjeyd commented:
@sarina Did all of the email addresses you enrolled via the membership tab belong to registered (existing) users?

itsjeyd commented:
@sarina For me the application behaves as follows:

If I go to the membership tab and enroll one or more students who are registered with the system, they are enrolled right away:

---

![existing-user-successfully-enrolled](https://cloud.githubusercontent.com/assets/961441/7791411/126e2274-02a0-11e5-8ced-0962eae6bb1d.png)

---

If I then generate a "may enroll" report for the course, it won't include users enrolled in the previous step because they are already members of the course.

On the other hand, if I enroll a number of students who are not registered with the system they will either be allowed to enroll after registration:

---

![non-existing-users-allowed-to-enroll](https://cloud.githubusercontent.com/assets/961441/7791488/5cd17dc4-02a1-11e5-9943-95224c18f3cf.png)

---

... or set to be automatically enrolled upon registration:

---

![non-existing-users-auto-enrolled-upon-registration](https://cloud.githubusercontent.com/assets/961441/7791505/a55d158a-02a1-11e5-8319-c06e9f8942ac.png)

---

Not being enrolled in the course, these users will be included in any "may enroll" report I generate afterwards.

Does this behavior differ from what you would expect to happen (and if so, how)?

Kelketek commented on common/djangoapps/student/models.py:
```diff
@@ -1293,6 +1293,23 @@ class Meta:  # pylint: disable=missing-docstring
+        may_enroll = CourseEnrollmentAllowed.objects.filter(course_id=course_id)
+        return (
+            student for student in may_enroll if not
+            enrolled.filter(email=student.email).exists()
```
This is going to require a bunch of database queries. Make use of a flat value set to simplify this. https://docs.djangoproject.com/en/1.4/ref/models/querysets/#values-list

Kelketek commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1031,6 +1031,29 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+    query_features = ['email']
+    try:
+        instructor_task.api.submit_calculate_may_enroll_csv(request, course_key, query_features)
+        success_status = _("Your students who may enroll report is being generated! You can view the status of the generation task in the 'Pending Instructor Tasks' section.")  # pylint: disable=line-too-long
```
Don't disable pylint messages that can be reasonably resolved. You can break the string into sections, as Python autoconcatinates strings between parens, allowing you to break this into multiple lines.

Kelketek commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1031,6 +1031,29 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+        success_status = _("Your students who may enroll report is being generated! You can view the status of the generation task in the 'Pending Instructor Tasks' section.")  # pylint: disable=line-too-long
+        return JsonResponse({"status": success_status})
+    except AlreadyRunningError:
+        already_running_status = _("A students who may enroll report generation task is already in progress. Check the 'Pending Instructor Tasks' table for the status of the task. When completed, the report will be available for download in the table below.")  # pylint: disable=line-too-long
```
Same note here.

itsjeyd commented:
@Kelketek Thanks for reviewing, I added two more commits that incorporate your feedback.

Kelketek commented:
Thanks, @itsjeyd . I've been trying to spin up an instance of your branch, but something appears to be wrong with manage.py's loading of the devstack lms both on yours and the master (I imagine you rebased atop master after whatever caused this to happen), though not on the last release, so I'm tracking down that so I can review the front-end as well. I'll look at your changes soon. @sarina do you know anything about this?

itsjeyd commented:
@Kelketek Yes -- I did rebase on `master` before opening this PR, and I also started having issues with `manage.py` after that. I bisected the commit range between the original merge base of my branch and `master`, and found that the first bad commit is 4312c0e.

The following patch to `manage.py` fixes the issue for me:

    diff --git a/manage.py b/manage.py
    index 266e656..c14d386 100755
    --- a/manage.py
    +++ b/manage.py
    @@ -113,5 +113,4 @@ if __name__ == "__main__":
     
         from django.core.management import execute_from_command_line
     
    -    sys.argv[1:] = django_args
    -    execute_from_command_line(sys.argv)
    +    execute_from_command_line([sys.argv[0]] + django_args)

Kelketek commented:
@itsjeyd Confirmed, this works for me. That's rather odd. The git blame for those lines goes all the way back to April, so I didn't think it could be the cause. But I think the commit was part of a more recent merge, so the date was deceptive. Going to update your branch and look at your work now.

Kelketek commented:
@itsjeyd This PR does work for me (the extended instructions you gave to @sarina helped me reason through it better to test). I've got that one little note remaining. You've got a new quality issue, but if you address that note, it should vanish, since it replaces the return you're currenty using. Once you've have that done, it's :+1: from me.

itsjeyd commented:
@Kelketek Done. Thanks again for reviewing!

antoviaque commented:
FYI the issue with the paver command seem to come from https://github.com/edx/edx-platform/commit/4312c0e763658c4326da8e1f4ce93b48b977160c#commitcomment-11347925

mhoeber commented:
DOC-2023 and DOC-2024

itsjeyd commented:
@sarina Did you get a chance to look at my [comment](https://github.com/edx/edx-platform/pull/8167#issuecomment-105121838) about the issue that you [raised](https://github.com/edx/edx-platform/pull/8167#issuecomment-104737243) on Friday? Please let me know if there's anything left you would like me to change.

Should I squash my commits now so you can have another look at the entire changeset?

mhoeber commented on lms/templates/instructor/instructor_dashboard_2/data_download.html:
```diff
@@ -31,6 +31,10 @@
 
     <p><input type="button" name="list-profiles-csv" value="${_("Download profile information as a CSV")}" data-endpoint="${ section_data['get_students_features_url'] }" data-csv="true"></p>
 
+    <p>${_("Click to generate a CSV file of all students who may enroll in this course but who have not yet signed up for it:")}</p>
```
Change to:  "Click to generate a CSV file that lists learners who can enroll in the course but have not yet done so."

itsjeyd replied:
Addressed in 7bd2a12.

mhoeber commented on lms/templates/instructor/instructor_dashboard_2/data_download.html:
```diff
@@ -31,6 +31,10 @@
 
+    <p>${_("Click to generate a CSV file of all students who may enroll in this course but who have not yet signed up for it:")}</p>
+
+    <p><input type="button" name="list-may-enroll-csv" value="${_("Download info about students who may enroll as a CSV")}" data-endpoint="${ section_data['get_students_who_may_enroll_url'] }" data-csv="true"></p>
```
Change to:  "Download a CSV of learners who can enroll"

itsjeyd replied:
Addressed in 7bd2a12.

mhoeber commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1031,6 +1031,36 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+    try:
+        instructor_task.api.submit_calculate_may_enroll_csv(request, course_key, query_features)
+        success_status = _(
+            "Your students who may enroll report is being generated! "
```
Change to: "The report of learners who can enroll is being generated."

itsjeyd replied:
@mhoeber This is a breaking change.

It affects the following test: [`test_calculate_report_csv_success`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/tests/test_api.py#L2276)

The test breaks because it requires success messages for CSV generation tasks to conform to the following format:

    Your {report_type} report is being generated! 
    You can view the status of the generation task in the 'Pending Instructor Tasks' section.

Currently, all of the CSV generation tasks produce success messages that conform to this format:

* Grade reports:

        Your grade report is being generated!
        You can view the status of the generation task in the 'Pending Instructor Tasks' section.

* Profile information for enrolled students:

        Your enrolled student profile report is being generated! 
        You can view the status of the generation task in the 'Pending Instructor Tasks' section.

* List of students who may enroll:

        Your students who may enroll report is being generated! 
        You can view the status of the generation task in the 'Pending Instructor Tasks' section.

We could of course relax the test to make sure it still passes after applying the changes you requested. However, I think it would be preferable to keep enforcing consistency between success messages of different tasks. One way of achieving this would be to change the success messages for *all* of the CSV generation tasks.

CC @sarina

mhoeber replied:
@itsjeyd  thanks I didn't consider that.  Outside the scope of this PR so ignore.

mhoeber commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1031,6 +1031,36 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+        instructor_task.api.submit_calculate_may_enroll_csv(request, course_key, query_features)
+        success_status = _(
+            "Your students who may enroll report is being generated! "
+            "You can view the status of the generation task in the 'Pending Instructor Tasks' section."
```
Change to:  "View the report status in the 'Pending Instructor Tasks' section."

itsjeyd replied:
@mhoeber This is a breaking change, cf. [this comment](https://github.com/edx/edx-platform/pull/8167#discussion_r31419660).

CC @sarina

mhoeber commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1031,6 +1031,36 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+        already_running_status = _(
+            "A students who may enroll report generation task is already in progress. "
+            "Check the 'Pending Instructor Tasks' table for the status of the task. "
+            "When completed, the report will be available for download in the table below."
```
Change to:
"Report generation is in progress."
"View the report status in the 'Pending Instructor Tasks' section."
"When the report is ready, download it from the table below."

itsjeyd replied:
@mhoeber This is a breaking change.

It affects the following test: [`test_calculate_report_csv_already_running`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/tests/test_api.py#L2289)

The test breaks because it requires "already running" messages for CSV generation tasks to conform to the following format:

    {report_type} report generation task is already in progress. 
    Check the 'Pending Instructor Tasks' table for the status of the task. 
    When completed, the report will be available for download in the table below.

Currently, all of the CSV generation tasks produce "already running" messages that conform to this format:

* Grade reports:

        A grade report generation task is already in progress. 
        Check the 'Pending Instructor Tasks' table for the status of the task. 
        When completed, the report will be available for download in the table below.

* Profile information for enrolled students:

        An enrolled student profile report generation task is already in progress.
        Check the 'Pending Instructor Tasks' table for the status of the task.
        When completed, the report will be available for download in the table below.

* List of students who may enroll:

        A students who may enroll report generation task is already in progress. 
        Check the 'Pending Instructor Tasks' table for the status of the task. 
        When completed, the report will be available for download in the table below.

We could of course relax the test to make sure it still passes after applying the changes you requested. However, I think it would be preferable to keep enforcing consistency between "already running" messages of different tasks. One way of achieving this would be to change the "already running" messages for *all* of the CSV generation tasks.

CC @sarina

mhoeber commented:
String comments added.

itsjeyd commented:
@mhoeber Thanks for the comments. I already changed the strings in `data_download.html`. Changing the strings in `instructor/views/api.py` breaks two existing tests:

* [`test_calculate_report_csv_success`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/tests/test_api.py#L2276)

* [`test_calculate_report_csv_already_running`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/tests/test_api.py#L2276)

This is because these tests check for exact wording in the success/failure messages returned by different methods that respond to requests for CSV files (cf. last two lines of each test).

Making these tests pass would require changing not only the assertions that they make but also the success/failure messages returned by the [`calculate_grades_csv`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/views/api.py#L2003) and [`get_students_features`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/views/api.py#L990) functions. These functions provide functionality that is not directly related to this PR.

How do you normally handle situations like this?

CC @sarina @antoviaque

sarina commented:
@itsjeyd Hi, and sorry for the late response; it was a busy & short week last week.

Generally, we work with the documentation folk to understand how necessary the requested string change is (ie, "must" vs "nice to have"). If the change is a "must", we change all the tests associated with it - especially since it's annoying but not too much work. Could you reply to @mhoeber 's comments that request breaking changes? Pointing out where the string is used elsewhere can give Mark enough context to understand whether or not the change is a "must". 

Also, this PR will require a rebase.

sarina commented:
@itsjeyd you asked:
> Did all of the email addresses you enrolled via the membership tab belong to registered (existing) users?

No. I registered 4 email addresses that had no account. I would expect those to show up in the report.

Tomorrow I will spin up a new sandbox and run another test.

sarina commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -2022,6 +2035,30 @@ def test_get_students_features_cohorted(self, is_cohorted):
+        """
+        url = reverse(
+            'get_students_who_may_enroll',
+            kwargs={'course_id': self.course.id.to_deprecated_string()}
```
This is old syntax; we're migrating to `unicode(course_id)`

itsjeyd replied:
Addressed in 734205c.

sarina commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -553,6 +557,54 @@ def test_unicode_usernames(self, students):
+        self.assertEquals(len(links), 1)
+        self.assertDictContainsSubset({'attempted': 1, 'succeeded': 1, 'failed': 0}, result)
+
+    @ddt.data([u'student@example.com', u'ni\xf1o@example.com'])
```
I don't get why you're using ddt for a test that's only called once

itsjeyd replied:
Addressed in fddeda3.

sarina commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -553,6 +557,54 @@ def test_unicode_usernames(self, students):
+            self._create_enrollment(email)
+
+        self.current_task = Mock()  # pylint: disable=attribute-defined-outside-init
+        self.current_task.update_state = Mock()
```
Not sure why you define these, nothing is asserted about them

itsjeyd replied:
Addressed in d8a488e.

sarina commented on lms/templates/courseware/legacy_instructor_dashboard.html:
```diff
@@ -329,8 +329,14 @@ <h2 class="navbar">[ <a href="#" onclick="goto('Grades');" class="${modeflag.get
-    <input type="submit" name="action" value="List enrolled students" class="${'is-disabled' if disable_buttons else ''}" aria-disabled="${'true' if disable_buttons else 'false'}">
-    <input type="submit" name="action" value="List students who may enroll but may not have yet signed up" class="${'is-disabled' if disable_buttons else ''}" aria-disabled="${'true' if disable_buttons else 'false'}" >
+    <p>
+      <input type="submit" name="action" value="List enrolled students" class="${'is-disabled' if disable_buttons else ''}" aria-disabled="${'true' if disable_buttons else 'false'}">
```
As per the discussion on https://openedx.atlassian.net/browse/OSPR-171 we're not planning to port this. Can you please add a message that direct users to the Data Download section's feature to "Download Profile Information as a CSV"

itsjeyd replied:
Addressed in 71d2574.

Screenshot:

![legacy-dashboard-button-removed](https://cloud.githubusercontent.com/assets/961441/7910522/ce4e18b8-0855-11e5-847b-dc6e8af2f4be.png)

sarina commented:
@itsjeyd please respond to @mhoeber's comments (either address them, or add comments explaining what the change would break and clarifying if the change should be made). I wrote a few comments as well. I'll make a sandbox and do functionality testing tomorrow.

itsjeyd commented:
@sarina Thanks for the info about string review. I addressed/responded to all of @mhoeber's comments individually, and also addressed your latest comments. I'll squash and rebase once you're done reviewing functionality.

sarina commented:
@itsjeyd I think last time I may have been looking at the wrong CSV files. I just retested, and this looks good to me! :+1: 

I will make sure Mark looks at this as soon as he can to answer your last remaining questions.

itsjeyd commented:
@sarina That's great, thanks a lot!

sarina commented:
@itsjeyd - you don't have to make the breaking changes, w00t!

Please squash and rebase this branch. Also, please add yourself to AUTHORS. Ping me when you've done that and I will merge as fast as I can (want to avoid there being a conflict in AUTHORS). Alternatively feel free to make a second pr just to add yourself to AUTHORS so you don't have to deal with potential rebasing headache.

itsjeyd commented:
> Please squash and rebase this branch. Also, please add yourself to AUTHORS.

@sarina Done and done. PING! :)

itsjeyd commented:
@sarina There is a single lettuce test that's failing in the build:

> LMS.Answer problems : I can reset a non-randomized problem that I answer incorrectly.| never | incorrect | drop down | (from lettuce)

Stacktrace:

    Traceback (most recent call last):
      File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/lettuce/core.py", line 144, in __call__
        ret = self.function(self.step, *args, **kw)
      File "/home/jenkins/workspace/edx-platform-test-subset/lms/djangoapps/courseware/features/problems.py", line 184, in assert_answer_mark
        assert(has_expected)
    AssertionError

This seems to be completely unrelated to the changes in this PR. What do you suggest?

sarina commented:
@itsjeyd we have a flaky test list here: bit.ly/1w91U41 Generally I suggest you go through that list and see if your failing test appears on it. If so, then you know it's flaky. If not, run another build and if that build passes, file a ticket marking the failing test as flaky.

However, you'll need to do another rebase anyhow for conflicts, so do that and a new build will be triggered.

itsjeyd commented:
@sarina Thanks, I checked the list and the test that failed is on it ([ticket](https://openedx.atlassian.net/browse/TE-572)).

sarina commented:
@itsjeyd - it's a good idea to make a comment on a ticket when you hit a flaky test to indicate that it's still an issue. I'll make a comment now.

itsjeyd commented:
@sarina OK, noted :) Also, the new build just passed!

sarina commented:
@itsjeyd ugh this was good to merge but I was out of the office, then my guess is someone, yet again, made an addition to AUTHORS. Wanna just remove that addition and make it in a separate PR, because the build won't matter against AUTHORS and I can just merge that quickly.

itsjeyd commented:
@sarina OK, I removed the commit that adds me to AUTHORS. Thanks for merging that change via a separate PR!

itsjeyd commented:
@sarina ... and I also rebased. We should be good to go!

sarina commented:
:confetti_ball:

lamagnifica commented:
@itsjeyd or @sarina, will interaction with this new button on the Instructor Dash result in an event being emitted? See http://edx.readthedocs.org/projects/devdata/en/latest/internal_data_formats/tracking_logs.html#instructor-events for a list of events emitted by some similar interactions.

sarina commented:
I don't think so. I was unaware we emitted events for instructor actions.

sarina commented:
@lamagnifica actually you should double check those events with the Analytics team. Looking through the platform, the only place I can find most of those strings are from the **legacy** dashboard. I don't think the instructor dash emits any events. Eg

```
lms/djangoapps/instructor/views/legacy.py:202:        track.views.server_track(request, "dump-answer-dist-csv", {}, page="idashboard")

lms/djangoapps/instructor/views/legacy.py:194:        track.views.server_track(request, "dump-grades-raw", {}, page="idashboard")

lms/djangoapps/instructor/views/legacy.py:197:        track.views.server_track(request, "dump-grades-csv-raw", {}, page="idashboard")
```

Some strings I can't find anywhere in the platform, such as `dump-graded-assignments-config` - which makes sense, we deprecated and removed that code from the legacy dash.

lamagnifica commented:
ooh, thanks @sarina!
